sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnnecessaryImports.cs

```diff
@@ -0,0 +1,207 @@
+                    {
+                        visitor.SafeVisit(member);
+                    }
+                    foreach (var attribute in compilationUnit.AttributeLists)
+                    {
+                        visitor.SafeVisit(attribute);
+                    }
+
+                    CheckDuplicateUsings(c, new HashSet<EquivalentNameSyntax>(), simpleNamespaces);
+                    CheckUnnecessaryUsings(c, simpleNamespaces, visitor.usedNamespaces);
+                },
+                SyntaxKind.CompilationUnit);
+
+        }
+
+        private static void CheckDuplicateUsings(SyntaxNodeAnalysisContext context, HashSet<EquivalentNameSyntax> ancestorsUsingDirectives, IEnumerable<UsingDirectiveSyntax> usingDirectives)
+        {
+            var groupingDirectives = usingDirectives
+                .GroupBy(usingDirective => new EquivalentNameSyntax(usingDirective.Name))
+                .ToList();
+
+            foreach (var potentialDuplicate in groupingDirectives)
+            {
+                var duplicates = potentialDuplicate.Skip(1);
+                foreach (var duplicate in duplicates)
+                {
+                    context.ReportDiagnosticWhenActive(Diagnostic.Create(rule, duplicate.GetLocation(), "duplicate"));
+                }
+
+                if (ancestorsUsingDirectives.Contains(potentialDuplicate.Key))
+                {
+                    context.ReportDiagnosticWhenActive(Diagnostic.Create(rule, potentialDuplicate.First().GetLocation(), "duplicate"));
+                }
+            }
+        }
+
+        private static void CheckUnnecessaryUsings(SyntaxNodeAnalysisContext context, IEnumerable<UsingDirectiveSyntax> usingDirectives, HashSet<INamespaceSymbol> usedNamespaces)
+        {
+            foreach (var usingDirective in usingDirectives)
+            {
+                if (context.SemanticModel.GetSymbolInfo(usingDirective.Name).Symbol is INamespaceSymbol namespaceSymbol
+                && !usedNamespaces.Any(usedNamespace => IsSameNamespace(usedNamespace, namespaceSymbol)))
+                {
+                    context.ReportDiagnosticWhenActive(Diagnostic.Create(rule, usingDirective.GetLocation(), "unnecessary"));
+                }
+            }
+        }
+
+        private static bool IsSameNamespace(INamespaceSymbol namespace1, INamespaceSymbol namespace2)
+        {
+            return namespace1.IsGlobalNamespace && namespace2.IsGlobalNamespace
+                || (namespace1.Name.Equals(namespace2.Name)
+                && namespace1.ContainingNamespace != null
+                && namespace2.ContainingNamespace != null
+                && IsSameNamespace(namespace1.ContainingNamespace, namespace2.ContainingNamespace));
+        }
+
+        private class CSharpRemovableUsingWalker : CSharpSyntaxWalker
+        {
+            private readonly SyntaxNodeAnalysisContext context;
+            private readonly HashSet<EquivalentNameSyntax> usingDirectivesFromParent;
```

Comment:
Its not really global using directives. When going inside nested namespace, its the sum of all parent using directives.
But it can indeed be made immutable, as once we're inside a specific scope, it won't change